hub/api/dataset.py

```diff
@@ -60,8 +59,10 @@ def __init__(
         self.storage = generate_chain(
             base_storage, memory_cache_size_bytes, local_cache_size_bytes, path
         )
-        self.tensors: Dict[str, Tensor] = {}
+        if self.mode == "r":
+            self.storage.enable_readonly()
```

Comment:
can we have `MODES=["r", "a"]` be a global here and raise an exception when a user provides an invalid mode?

further, it would be nice to has `ds.mode = "r"` work, a nice design here would be to have the `@mode.setter` syntax here, and within the `Dataset.__init__` you can call `self.mode = "mode"` to trigger these validations

Reply:
then, `storage.disable_readonly()` when `ds.mode = "w"` is called

Reply:
or, better yet `storage.mode = "w"` and `storage.mode = "r"` in the same way that is done in `hub.Dataset`.

up to you

Reply:
@McCrearyD As far as I understand, `mode='w'` will overwrite the contents. This doesn't really make sense to be changed dynamically: if you set from 'r' to 'w', what happens to the contents?

I'll add the mode check for now.

https://stackoverflow.com/questions/1466000/difference-between-modes-a-a-w-w-and-r-in-built-in-open-function